src/data/flamegraphs.rs

```diff
@@ -31,39 +40,76 @@ impl FlamegraphRaw {
+    fn prepare_data_collector(&mut self, _params: CollectorParams) -> Result<()> {
+        match Command::new("perf").args(["--version"]).output() {
+            Err(e) => Err(PDError::DependencyError(format!("'perf' command failed. {}", e)).into()),
+            _ => Ok(()),
+        }
+    }
+
     fn after_data_collection(&mut self, params: CollectorParams) -> Result<()> {
-        let data_dir = PathBuf::from(params.data_dir.clone());
+        let data_dir = PathBuf::from(&params.data_dir);
 
-        let mut file_pathbuf = data_dir.clone();
-        file_pathbuf.push(get_file_name(
-            params.data_dir.clone(),
-            "perf_profile".to_string(),
-        )?);
+        let file_pathbuf =
+            data_dir.join(get_file_name(params.data_dir, "perf_profile".to_string())?);
 
-        let mut perf_jit_loc = data_dir.clone();
-        perf_jit_loc.push(format!("{}-perf.data.jit", params.run_name));
+        let perf_jit_loc = data_dir.join("perf.data.jit");
 
-        println!("Running Perf inject...");
+        trace!("Running Perf inject...");
         let out_jit = Command::new("perf")
             .args([
                 "inject",
                 "-j",
                 "-i",
                 &file_pathbuf.to_str().unwrap(),
                 "-o",
-                perf_jit_loc.clone().to_str().unwrap(),
+                perf_jit_loc.to_str().unwrap(),
             ])
-            .status();
+            .output();
+
+        let fg_loc = data_dir.join(format!("{}-flamegraph.svg", params.run_name));
+        let mut fg_out = File::create(&fg_loc)?;
+
         match out_jit {
             Err(e) => {
-                if e.kind() == ErrorKind::NotFound {
-                    error!("'perf' command not found.");
-                } else {
-                    error!("Unknown error: {}", e);
+                let out = format!("Skip processing profiling data due to: {}", e);
+                error!("{}", out);
+                write_msg_to_svg(fg_out, out)?;
+            }
+            Ok(_) => {
+                trace!("Perf inject successful.");
+                let script_loc = data_dir.join("script.out");
+                let collapse_loc = data_dir.join("collapse.out");
+
+                info!("Creating flamegraph...");
+                let mut script_out = File::create(&script_loc)?;
+                let collapse_out = File::create(&collapse_loc)?;
+
+                let out = Command::new("perf")
+                    .args(["script", "-f", "-i", perf_jit_loc.to_str().unwrap()])
+                    .output();
+                match out {
+                    Err(e) => {
+                        let out = format!("Did not process profiling data due to: {}", e);
+                        error!("{}", out);
+                        write_msg_to_svg(fg_out, out)?;
+                    }
+                    Ok(v) => {
+                        write!(script_out, "{}", std::str::from_utf8(&v.stdout)?)?;
```

Comment:
Can you try to get this so that 'perf' is writing the file directly, versus slurping a potentially huge output up into this process, only to then write it out to a file?

Reply:
Yeah. We can set the `stdout` and give it the file and that should work.